engine/execution/provider/engine.go

```diff
@@ -24,150 +28,174 @@ import (
 		return nil, fmt.Errorf("could not register chunk data pack provider engine: %w", err)
 	}
-	eng.chunksConduit = chunksConduit
-
-	return &eng, nil
-}
+	engine.chunksConduit = chunksConduit
 
-func (e *Engine) SubmitLocal(event interface{}) {
-	e.unit.Launch(func() {
-		err := e.ProcessLocal(event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
+	cm := component.NewComponentManagerBuilder()
+	for i := uint(0); i < chdpRequestWorkers; i++ {
+		cm.AddWorker(engine.processChunkDataPackRequestWorker)
+	}
 
-func (e *Engine) Submit(channel channels.Channel, originID flow.Identifier, event interface{}) {
-	e.unit.Launch(func() {
-		err := e.Process(channel, originID, event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
+	engine.cm = cm.Build()
+	engine.Component = engine.cm
 
-func (e *Engine) ProcessLocal(event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(e.me.NodeID(), event)
-	})
+	return &engine, nil
 }
 
 // Ready returns a channel that will close when the engine has
 // successfully started.
 func (e *Engine) Ready() <-chan struct{} {
-	return e.unit.Ready()
+	return e.cm.Ready()
 }
 
 // Done returns a channel that will close when the engine has
 // successfully stopped.
 func (e *Engine) Done() <-chan struct{} {
-	return e.unit.Done()
+	return e.cm.Done()
+}
+
+// processChunkDataPackRequestWorker encapsulates the logic of a single (concurrent) worker that picks a chunk data pack
+// request from this engine's queue and processes it.
+func (e *Engine) processChunkDataPackRequestWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+	ticker := time.NewTicker(e.chdpRequestProcessInterval)
+	e.log.Debug().
+		Dur("process_inveral", e.chdpRequestProcessInterval).
+		Msg("process chunk data pack request worker started")
+
+	for {
+		select {
+		case <-ticker.C:
+			request, exists := e.chdpRequestsQueue.Pop()
+			if !exists {
+				e.log.Trace().Msg("worker did not find any chunk data pack request, sleeping till next cycle")
+				continue
+			}
+			lg := e.log.With().
+				Hex("chunk_id", logging.ID(request.ChunkId)).
+				Hex("origin_id", logging.ID(request.RequesterId)).Logger()
+			lg.Trace().Msg("chunk data pack request picked for processing")
+			e.onChunkDataRequest(request)
+			lg.Trace().Msg("chunk data pack request processed")
+		case <-ctx.Done():
+			e.log.Trace().Msg("processing chunk data pack request worker terminated")
+			return
+		}
+	}
 }
 
 func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+	return e.process(originID, event)
 }
 
 func (e *Engine) process(originID flow.Identifier, event interface{}) error {
-	ctx := context.Background()
 	switch v := event.(type) {
 	case *messages.ChunkDataRequest:
-		e.onChunkDataRequest(ctx, originID, v)
+		pushed := e.chdpRequestsQueue.Push(v.ChunkID, originID)
```

Comment:
so if push fails here, who is responsible for re-pushing it?

Reply:
Push only fails on two circumstances: (1) queue is full (2) the pair `v.ChunkID` and `originID` already exists in the queue. 

In case (1), the request will be repeated by the verification node after a backoff. In other words, when execution node is dropping a chunk data pack request, it means the node has reached its capacity on processing such requests. So when the verification node does not hear back about its chunk data pack request, it realizes something is wrong with the execution node, so the verification node by its protocol is backing off exponentially and retry the request a few seconds (up to a minute) later. 

In case (2), the request already exists in the cache and will literally be processed by the execution node, so there is no need to add a duplicate one.